src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,2059 @@
+                decoder_ngram_attentions=outputs.decoder_ngram_attentions,
+                decoder_cross_attentions=outputs.decoder_cross_attentions,
+                encoder_last_hidden_state=outputs.encoder_last_hidden_state,
+                encoder_hidden_states=outputs.encoder_hidden_states,
+                encoder_attentions=outputs.encoder_attentions,
+            )
+
+    def _compute_loss(self, logits, labels):
+        expend_targets = labels.new_zeros(self.config.ngram, labels.size(0), labels.size(1)).fill_(self.padding_idx)
+
+        for i in range(self.config.ngram):
+            if i > 0 and self.disable_ngram_loss:
+                break
+            expend_targets[i, :, :] = labels
+
+        lprobs = F.log_softmax(
+            logits.view(-1, logits.size(-1)),
+            dim=-1,
+            dtype=torch.float32,
+        )
+
+        loss = F.nll_loss(lprobs, expend_targets.view(-1), reduction="sum")
+
+        if self.config.eps > 0.0:
+            smooth_loss = -lprobs.sum(dim=-1, keepdim=True)
+            non_pad_mask = expend_targets.ne(self.padding_idx).view(-1)
+            smooth_loss = smooth_loss[non_pad_mask]
+            smooth_loss = smooth_loss.sum()
+
+            eps_i = self.config.eps / lprobs.size(-1)
+            loss = (1.0 - self.config.eps) * loss + eps_i * smooth_loss
+
+        return loss
+
+    def prepare_inputs_for_generation(
+        self, decoder_input_ids, past, attention_mask, use_cache, encoder_outputs, **kwargs
+    ):
+        assert encoder_outputs is not None, "`encoder_outputs` have to be passed for generation."
+
+        if past:
+            decoder_input_ids = decoder_input_ids[:, -1:]
+        # first step, decoder_cached_states are empty
+        return {
+            "input_ids": None,  # encoder_outputs is defined. input_ids not needed
+            "encoder_outputs": encoder_outputs,
+            "past_key_values": past,
+            "decoder_input_ids": decoder_input_ids,
+            "attention_mask": attention_mask,
+            "use_cache": use_cache,
+        }
+
+    def prepare_logits_for_generation(self, logits, cur_len, max_length):
+        if (cur_len == 1) or (cur_len == max_length - 1 and self.config.eos_token_id is not None):
+            return self._force_token_ids_generation(logits, self.config.bos_token_id)
+        return logits
+
+    def _force_token_ids_generation(self, scores, token_ids) -> None:
+        """force one of token_ids to be generated by setting prob of all other tokens to 0"""
+        if isinstance(token_ids, int):
+            token_ids = [token_ids]
+        all_but_token_ids_mask = torch.tensor(
+            [x for x in range(self.config.vocab_size) if x not in token_ids],
+            dtype=torch.long,
+            device=next(self.parameters()).device,
+        )
+        assert len(scores.shape) == 2, "Scores should be of rank 2"
+        scores[:, all_but_token_ids_mask] = -float("inf")
+        return scores
```

Comment:
```suggestion
    def adjust_logits_during_generation(self, logits, cur_len, max_length):
        if (cur_len == 1) or (cur_len == max_length - 1):
            self._force_token_id_to_be_generated(logits, self.config.bos_token_id)
        return logits
    def _force_token_id_to_be_generated(self, scores, token_id) -> None:
        """force one of token_ids to be generated by setting prob of all other tokens to 0 (logprob=-float("inf"))"""
        scores[:, [x for x in range(self.config.vocab_size) if x != token_id]] = -float("inf")
```

That code adjusts to new names, and simplifies since we are always forcing an int (not a list) to be generated.
+ moderately concerning that integration tests pass without that fix.
+ also weird that you are forcing `bos_token_id` when you hit `max_length`